Propagate stats and cost in distributed explain plans

rschlussel opened the pull request:
fixes #10963

kokosing commented on presto-main/src/main/java/com/facebook/presto/cost/FragmentedPlanSourceProvider.java:
```diff
@@ -0,0 +1,61 @@
+public class FragmentedPlanSourceProvider
+        implements PlanNodeSourceProvider
+{
+    private Map<PlanFragmentId, PlanNode> fragmentPlanNodeMap;
```
`planFragments`?

rschlussel replied:
Changing to planFragmentsMap.  I don't like when maps are named without the word map (or something that makes it clear it's a mapping).  I always expect them to be a list or a set (or any other collection of just one thing).

martint replied:
Heh... I actually dislike when variable names encode their type (unless there's a very good reason for that, such as a conflict with some other variable -- e.g., a when a builder is used, or when having to convert types and you need to hold on to both versions). In this particular case, fragmentPlanNodeMap doesn't even convey what it represents. I would suggest sourcesByFragment, nodesByFragment, sourceNodesByFragment or something like that.

kokosing commented on presto-main/src/main/java/com/facebook/presto/cost/FragmentedPlanSourceProvider.java:
```diff
@@ -0,0 +1,61 @@
+        return node.getSources();
+    }
+
+    public static FragmentedPlanSourceProvider createFragmentedPlanSourceProvider(List<PlanFragment> planFragments)
```
`s/createFragmentedPlanSourceProvider/create`

Also I would put it at the top, above the constructor

kokosing commented on presto-main/src/main/java/com/facebook/presto/cost/FragmentedPlanStatsCalculator.java:
```diff
@@ -0,0 +1,92 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
```
double licence

kokosing commented on presto-main/src/main/java/com/facebook/presto/cost/FragmentedPlanStatsCalculator.java:
```diff
@@ -0,0 +1,92 @@
+    public FragmentedPlanStatsCalculator(StatsCalculator delegate, FragmentedPlanSourceProvider sourceProvider)
+    {
+        this.delegate = delegate;
+        this.sourceProvider = sourceProvider;
```
rnn

kokosing commented on presto-main/src/main/java/com/facebook/presto/cost/FragmentedPlanStatsCalculator.java:
```diff
@@ -0,0 +1,92 @@
+        this.sourceProvider = sourceProvider;
+    }
+
+    public PlanNodeStatsEstimate calculateRemoteSourceStats(RemoteSourceNode node, StatsProvider statsProvider, Lookup lookup, Session session, TypeProvider types)
```
should not this method be added as a regular rule to `com.facebook.presto.cost.ComposableStatsCalculator`?

kokosing replied:
Why is this public?

kokosing commented on presto-main/src/main/java/com/facebook/presto/cost/FragmentedPlanStatsCalculator.java:
```diff
@@ -0,0 +1,92 @@
+    }
+
+    @Override
+    public PlanNodeStatsEstimate calculateStats(PlanNode node, StatsProvider sourceStats, Lookup lookup, Session session, TypeProvider types)
```
move it to the top, just below the constructor

kokosing commented on presto-main/src/main/java/com/facebook/presto/cost/PlanNodeSourceProvider.java:
```diff
@@ -0,0 +1,24 @@
+
+import java.util.List;
+
+public interface PlanNodeSourceProvider
```
add `@FunctionalInterface`

kokosing commented on presto-main/src/main/java/com/facebook/presto/cost/FragmentedPlanStatsCalculator.java:
```diff
@@ -0,0 +1,92 @@
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Verify.verify;
+
+public class FragmentedPlanStatsCalculator
```
RemoteSourceStatsCalculator?

rschlussel replied:
it's a statscalculator for the whole fragmented plan.  it just handles a remote source node.

kokosing commented on presto-main/src/main/java/com/facebook/presto/cost/FragmentedPlanStatsCalculator.java:
```diff
@@ -0,0 +1,92 @@
+    private final StatsCalculator delegate;
+    private final FragmentedPlanSourceProvider sourceProvider;
+
+    public FragmentedPlanStatsCalculator(StatsCalculator delegate, FragmentedPlanSourceProvider sourceProvider)
```
Inject `PlanSourceProvider`

rschlussel replied:
It needs to be created new for each query.

kokosing commented on presto-main/src/main/java/com/facebook/presto/cost/FragmentedPlanStatsCalculator.java:
```diff
@@ -0,0 +1,92 @@
+import static com.google.common.base.Verify.verify;
+
+public class FragmentedPlanStatsCalculator
+        implements StatsCalculator
```
No tests for this class?

rschlussel replied:
I'll work on that tomorrow

rschlussel commented on presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/RuleAssert.java:
```diff
@@ -185,7 +185,7 @@ private RuleApplication applyRule()
     private String formatPlan(PlanNode plan, TypeProvider types)
     {
-        return inTransaction(session -> textLogicalPlan(plan, types, metadata.getFunctionRegistry(), statsCalculator, costCalculator, session, 2));
+        return inTransaction(session -> textLogicalPlan(plan, types, metadata.getFunctionRegistry(), statsCalculator, costCalculator, session));
```
yeah but it was equally readable as zero, so it didn't seem worth having its own specialized constructor.

rschlussel commented on presto-main/src/main/java/com/facebook/presto/cost/FragmentedPlanStatsCalculator.java:
```diff
@@ -0,0 +1,92 @@
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Verify.verify;
+
+public class FragmentedPlanStatsCalculator
```
I don't think we still need support for the `CoefficientBasedStatsCalculator`, and getting rid of the `SelectingStatsCalculator` would be nice too.  But unless we want to get rid of the StatsCalculator interface so that the only kind of StatsCalculator is a ComposableStatsCalculator, it can't be a rule. Honestly I don't think that's a bad idea because the ComposableStatsCalculator is by definition super extensible and flexible. What do you think?

arhimondr replied:
Yeah. I think we can remove the `CoefficientBasedStatsCalculator`, and rename `ComposableStatsCalculator` to `StatsCalculator` removing the interface.

kokosing replied:
https://github.com/prestodb/presto/pull/11286

rschlussel commented:
Addressed comments except writing tests and updated how PlanFragmentStatsCalculator gets the input symbols to fix the failing explain analyze tests.

kokosing commented on presto-main/src/main/java/com/facebook/presto/cost/FragmentedPlanStatsCalculator.java:
```diff
@@ -0,0 +1,83 @@
+        for (PlanFragment sourceFragment : sourceProvider.getSourceFragments(node)) {
+            PlanNodeStatsEstimate sourceStatsWithMappedSymbols = mapToOutputSymbols(statsProvider.getStats(sourceFragment.getRoot()), sourceFragment.getPartitioningScheme().getOutputLayout(), node.getOutputSymbols());
+
+            if (estimate != null) {
```
how estimate can become null?

kokosing commented on presto-main/src/main/java/com/facebook/presto/cost/FragmentedPlanStatsCalculator.java:
```diff
@@ -0,0 +1,83 @@
+    {
+        PlanNodeStatsEstimate estimate = null;
+        for (PlanFragment sourceFragment : sourceProvider.getSourceFragments(node)) {
+            PlanNodeStatsEstimate sourceStatsWithMappedSymbols = mapToOutputSymbols(statsProvider.getStats(sourceFragment.getRoot()), sourceFragment.getPartitioningScheme().getOutputLayout(), node.getOutputSymbols());
```
Maybe `FragmentedPlanSourceProvider::getSources` for `RemoteSourceNode` should return nodes with output symbols already mapped? That way you wouldn't need to map symbols here.

rschlussel replied:
i think that would mess up getting the stats for those sources.

findepi commented on presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorUsingExchanges.java:
```diff
@@ -238,6 +239,12 @@ public PlanNodeCostEstimate visitLimit(LimitNode node, Void context)
+        @Override
+        public PlanNodeCostEstimate visitRemoteSource(RemoteSourceNode node, Void context)
+        {
+            return ZERO_COST;
```
Add a comment here explaining why remote sources are ignored in cost estimation.

of course, throwing would be safer here, in case cost calculator is used on actual plan somewhere else than in EXPLAIN.

rschlussel replied:
I was just making this a pass through so we could get the costs for other things, but now I'm thinking that we'll lose the exchange cost. Maybe I should make a FragmentCostCalculator extension to get the exchange properties from the fragment and then call `calculateExchangeCost`

findepi replied:
What about different approach -- that we de-fragment the plan and pass that to the cost calculator? Then the stats & costing code would not need to be modified.

findepi replied:
> I was just making this a pass through so we could get the costs for other things, but now I'm thinking that we'll lose the exchange cost.

I didn't notice. The problem is that it's possible for "cost on fragmented plan" to be very off, and it may go unnoticed for long time: are we able to test-cover everything? will we notice in production when costs are off? (ultimately we will, but not too soon, as the costing here won't have any impact on execution)

rschlussel replied:
I had looked at defragmenting and computing the cost, but it gets very messy in the plan printer because each node you visit needs to have the defragmented plan and match up correctly (or re.-defragment), etc. 

I'm going to go with the FragmentedPlanCostCalculator approach for handling RemoteSourceNodes and I'll add a test that creates a plan and computes the cost and then fragments the plan and checks that the cumulative cost is the same.

findepi commented on presto-main/src/main/java/com/facebook/presto/cost/CachingCostProvider.java:
```diff
@@ -95,7 +102,7 @@ private PlanNodeCostEstimate calculateCumulativeCost(PlanNode node)
         PlanNodeCostEstimate localCosts = costCalculator.calculateCost(node, statsProvider, lookup, session, types);
 
-        PlanNodeCostEstimate sourcesCost = node.getSources().stream()
+        PlanNodeCostEstimate sourcesCost = sourceProvider.getSources(node).stream()
```
Can sourceProvider be modelled with a lookup?

rschlussel replied:
not without abusing the interface :).  Lookups are for resolving GroupReferences to the nodes that are a part of that group.  A SourceProvider gives you the list of source nodes for a given node.

findepi commented on presto-main/src/main/java/com/facebook/presto/cost/CachingCostProvider.java:
```diff
@@ -38,22 +38,29 @@
+        this(costCalculator, statsProvider, Optional.empty(), noLookup(), session, types, sourceProvider);
+    }
+
+    public CachingCostProvider(CostCalculator costCalculator, StatsProvider statsProvider, Optional<Memo> memo, Lookup lookup, Session session, TypeProvider types, PlanNodeSourceProvider sourceProvider)
```
Did you check this PR for potential conflicts with https://github.com/prestodb/presto/pull/11267?

rschlussel replied:
doesn't overlap i don't think.

rschlussel commented:
Updated: 
1. Added a commit to include stats in plans printed from QueryMonitor.
2. @haozhun beat me at getting rid of all the overloaded methods in PlanPrinter, so from that commit I now just have a small simplification in formatFragment.  
3. Added support for computing "exchange costs" for remote source nodes (which was a bit of an invasive change) and added tests for the fragmented plan cost and stats calculators

rschlussel commented:
Also @arhimondr. i didn't get rid of the stats calculator interface because it's useful for testing.

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/analyzer/QueryExplainer.java:
```diff
@@ -102,6 +112,8 @@ public QueryExplainer(
         this.statsCalculator = requireNonNull(statsCalculator, "statsCalculator is null");
         this.costCalculator = requireNonNull(costCalculator, "costCalculator is null");
+        this.nodeManager = requireNonNull(nodeManager, "nodeManager is null");
+        this.nodeSchedulerConfig = requireNonNull(nodeSchedulerConfig);
```
message

arhimondr commented on presto-main/src/main/java/com/facebook/presto/testing/LocalQueryRunner.java:
```diff
@@ -827,6 +827,8 @@ public Plan createPlan(Session session, @Language("SQL") String sql, List<PlanOp
                 statsCalculator,
                 costCalculator,
+                nodeManager,
+                new NodeSchedulerConfig().setIncludeCoordinator(true),
```
This config is being created twice in this class. Create in in the `LocalQueryRunner` constructor and store it as a field, to make sure it is always the same.

arhimondr commented on presto-main/src/test/java/com/facebook/presto/cost/TestCostCalculator.java:
```diff
@@ -74,7 +120,46 @@
     private final CostCalculator costCalculatorUsingExchanges = new CostCalculatorUsingExchanges(() -> NUMBER_OF_NODES);
     private final CostCalculator costCalculatorWithEstimatedExchanges = new CostCalculatorWithEstimatedExchanges(costCalculatorUsingExchanges, () -> NUMBER_OF_NODES);
-    private Session session = testSessionBuilder().build();
+    private final PlanFragmenter planFragmenter = new PlanFragmenter(new QueryManagerConfig());
```
- Nullify all the fields after test class is finished
- Use `createTestTransactionManager(catalogManager)`
- Add constructor method  that accepts `transactionManager` in `MetadataManager`, and use `createTestMetadataManager(new FeaturesConfig(), transactionManager)`
- Optionally you can create similar `createTest*` methods for `nodePartitioningManager`, that creates test `nodePartitioningManager` that isn't required to be stopped, and can be simply nullified (uses direct executors instead of thread pools, etc.)

Here is the example:

```
private CostCalculator costCalculatorUsingExchanges;
    private CostCalculator costCalculatorWithEstimatedExchanges;
    private PlanFragmenter planFragmenter;
    private Session session;
    private MetadataManager metadata;
    private TransactionManager transactionManager;
    private FinalizerService finalizerService;
    private NodeScheduler nodeScheduler;
    private NodePartitioningManager nodePartitioningManager;

    @BeforeClass
    public void setUp()
    {
        costCalculatorUsingExchanges = new CostCalculatorUsingExchanges(() -> NUMBER_OF_NODES);
        costCalculatorWithEstimatedExchanges = new CostCalculatorWithEstimatedExchanges(costCalculatorUsingExchanges, () -> NUMBER_OF_NODES);
        planFragmenter = new PlanFragmenter(new QueryManagerConfig());

        session = testSessionBuilder().setCatalog("tpch").build();

        CatalogManager catalogManager = new CatalogManager();
        catalogManager.registerCatalog(createBogusTestingCatalog("tpch"));
        transactionManager = createTestTransactionManager(catalogManager);
        metadata = createTestMetadataManager(new FeaturesConfig(), transactionManager);

        finalizerService = new FinalizerService();
        finalizerService.start();
        nodeScheduler = new NodeScheduler(
                new LegacyNetworkTopology(),
                new InMemoryNodeManager(),
                new NodeSchedulerConfig().setIncludeCoordinator(true),
                new NodeTaskMap(finalizerService));
        nodePartitioningManager = new NodePartitioningManager(nodeScheduler);
    }

    @AfterClass(alwaysRun = true)
    public void tearDown()
    {
        costCalculatorUsingExchanges = null;
        costCalculatorWithEstimatedExchanges = null;
        planFragmenter = null;
        session = null;
        transactionManager = null;
        metadata = null;
        finalizerService.destroy();
        finalizerService = null;
        nodeScheduler.stop();
        nodeScheduler = null;
        nodePartitioningManager = null;
    }
```

arhimondr commented on presto-main/src/test/java/com/facebook/presto/cost/TestFragmentedPlanStatsCalculator.java:
```diff
@@ -0,0 +1,199 @@
+
+public class TestFragmentedPlanStatsCalculator
+{
+    Metadata metadata = createQueryRunner().getMetadata();
```
Similar comment about bootstraping the test class

```
private Metadata metadata;
    private Session session;

    @BeforeMethod
    public void setUp()
    {
        metadata = createTestMetadataManager();
        session = testSessionBuilder().build();
    }

    @AfterMethod(alwaysRun = true)
    public void tearDown()
    {
        metadata = null;
        session = null;
    }
```

arhimondr commented on presto-main/src/test/java/com/facebook/presto/cost/TestFragmentedPlanStatsCalculator.java:
```diff
@@ -0,0 +1,199 @@
+                requireNonNull(stats.get(node), "no stats for node");
+    }
+
+    private static LocalQueryRunner createQueryRunner()
```
Remove this method

rschlussel commented:
Thanks for the review!  Updated and I'll merge on Monday.

wenleix commented on presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java:
```diff
@@ -274,14 +284,15 @@ private static String formatFragment(FunctionRegistry functionRegistry, StatsCal
-            builder.append(textLogicalPlan(fragment.getRoot(), TypeProvider.copyOf(fragment.getSymbols()), Optional.of(fragment.getStageExecutionStrategy()), functionRegistry, statsCalculator, costCalculator, session, Optional.empty(), 1, verbose))
-                    .append("\n");
-        }
+        TypeProvider typeProvider = TypeProvider.copyOf(allFragments.stream()
```
@rschlussel : My understanding is we need `allFragments` as we want to create `sourceProvider` and `statsCalculator`. However, later you make stats and cost precomputed and store in `PlanFragment`: https://github.com/prestodb/presto/pull/11511

I am wondering do we still need to instantiate the `TypeProvider` with all fragments?
